src/laws/Cache.jl

```diff
@@ -0,0 +1,68 @@
+import Base.similar, Base.size, Base.fill
+export MatrixCacheNoVJP, FloatCacheNoVJP, MatrixCache, FloatCache
+
+"""
+    MatrixCacheNoVJP
+
+A mutable cache structure for storing a two-dimensional array of `Float64` values.
+This struct is intended for use cases where the law is not differentiated, and hence the vector-Jacobian products (VJP) are not required.
+Fields:
+- `value::Array{Float64, 2}`: The cached matrix.
+"""
+mutable struct MatrixCacheNoVJP
+    value::Array{Float64, 2}
+end
+
+"""
+    FloatCacheNoVJP
+
+A mutable cache structure for storing a scalar value as a zero-dimensional array of `Float64`.
+This struct is intended for use cases where the law is not differentiated, and hence the vector-Jacobian products (VJP) are not required.
+Fields:
+- `value::Array{Float64, 0}`: The cached scalar value.
+"""
+mutable struct FloatCacheNoVJP
+    value::Array{Float64, 0}
+end
+
+similar(c::Union{FloatCacheNoVJP, MatrixCacheNoVJP}) = typeof(c)(similar(c.value))
+size(c::Union{FloatCacheNoVJP, MatrixCacheNoVJP}) = size(c.value)
+fill(c::Union{FloatCacheNoVJP, MatrixCacheNoVJP}, s) = typeof(c)(fill(c.value, s))
+Base.:(==)(a::Union{FloatCacheNoVJP, MatrixCacheNoVJP}, b::Union{FloatCacheNoVJP, MatrixCacheNoVJP}) = a.value == b.value
+
+"""
+    MatrixCache
+
+A cache structure for storing a two-dimensional array of `Float64` values along with
+their associated vector-Jacobian products (VJP).
+Fields:
+- `value::Array{Float64, 2}`: The cached matrix.
+- `vjp_inp::Array{Float64, 2}`: VJP with respect to inputs.
+- `vjp_θ::Vector{Float64}`: VJP with respect to parameters.
+"""
+struct MatrixCache
+    value::Array{Float64, 2}
+    vjp_inp::Array{Float64, 2}
+    vjp_θ::Vector{Float64}
+end
+
+"""
+    FloatCache
+
+A cache structure for storing a scalar value as a zero-dimensional array of `Float64` along with
+their associated vector-Jacobian products (VJP).
+Fields:
+- `value::Array{Float64, 0}`: The cached scalar value.
+- `vjp_inp::Array{Float64, 0}`: VJP with respect to inputs.
+- `vjp_θ::Vector{Float64}`: VJP with respect to parameters.
+"""
+struct FloatCache
+    value::Array{Float64, 0}
+    vjp_inp::Array{Float64, 0}
+    vjp_θ::Vector{Float64}
+end
+
```

Comment:
To understand the logic of this: depending the type of the inversion we will define a new cache? For example, in the case of the full D inversion, I can imagine defining a `InterpolationCache` that stores the value of the VJPs in a few nodes and then define a function that evaluates the VJP at every given point based on the cache values. Is this a possible implementation of the interpolation @albangossard ?